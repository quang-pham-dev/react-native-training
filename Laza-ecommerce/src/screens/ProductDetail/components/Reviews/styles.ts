import { StyleSheet } from 'react-native';

// Themes
import Colors from 'themes/Colors';
import Fonts from 'themes/Fonts';
import Metrics from 'themes/Metrics';

const styles = StyleSheet.create({
  reviewContainer: {
    marginTop: Metrics.margin.lg,
  },

  reviewTitleWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },

  textReviewTitle: {
    fontSize: Fonts.size.default,
    fontFamily: Fonts.fontFamily.Inter_600SemiBold,
    lineHeight: Fonts.lineHeight.sm,
    color: Colors.textBlack,
  },

  reviewWrapper: {
    width: '100%',
    justifyContent: 'space-between',
  },

  reviewItemWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: Metrics.margin.xm,
  },

  reviewerInfoWrapper: {
    flexDirection: 'row',
    alignItems: 'center',
  },

  imageAvatar: {
    width: 40,
    height: 40,
    borderRadius: 40,
  },

  reviewerItemInfo: {
    marginLeft: Metrics.margin.xm,
  },

  reviewTime: {
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: Metrics.margin.xs,
  },

  imageClock: {
    width: Metrics.icons.small,
    height: Metrics.icons.small,
  },

  textDate: {
    fontSize: Fonts.size.min,
    fontFamily: Fonts.fontFamily.Inter_500Medium,
    lineHeight: Fonts.lineHeight.xxs,
    color: Colors.textGray,
    marginLeft: Metrics.margin.xsm,
  },

  textName: {
    fontSize: Fonts.size.normal,
    fontFamily: Fonts.fontFamily.Inter_500Medium,
    lineHeight: Fonts.lineHeight.base,
  },
  reviewItemRating: {
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },

  ratingTextWrapper: {
    flexDirection: 'row',
    alignItems: 'center',
  },

  ratingTextPoint: {
    fontSize: Fonts.size.normal,
    fontFamily: Fonts.fontFamily.Inter_500Medium,
    lineHeight: Fonts.lineHeight.base,
    color: Colors.textBlack,
  },
  ratingText: {
    fontSize: Fonts.size.min,
    fontFamily: Fonts.fontFamily.Inter_400Regular,
    lineHeight: Fonts.lineHeight.xxs,
    color: Colors.textGray,
    marginLeft: Metrics.margin.xsm,
  },

  imageStar: {
    width: 57,
    height: 13,
  },
  reviewItemContent: {
    marginTop: Metrics.margin.sm,
  },
  textContent: {
    fontSize: Fonts.size.normal,
    fontFamily: Fonts.fontFamily.Inter_400Regular,
    lineHeight: Fonts.lineHeight.lg,
    color: Colors.textGray,
  },

  textTitle: {
    fontSize: Fonts.size.small,
    fontFamily: Fonts.fontFamily.Inter_400Regular,
    lineHeight: Fonts.lineHeight.xs,
    color: Colors.textGray,
  },
});

export default styles;
