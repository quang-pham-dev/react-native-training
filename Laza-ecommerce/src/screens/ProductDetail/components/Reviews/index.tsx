import React, { memo } from 'react';
import { View, Text, Image } from 'react-native';

// LIBS
import isEqual from 'react-fast-compare';

// Components
import Title from 'components/Title';
import MoreLessText from 'components/LessMoreText';

// Types
import { IProductDetailItemProps } from 'types/screens/ProductDetail';

// Themes
import IMAGES from 'themes/Images';

// Styles
import styles from './styles';

const Reviews = ({ comment, rating, reviewers }: IProductDetailItemProps) => {
  return (
    <View style={styles.reviewContainer}>
      <View style={styles.reviewTitleWrapper}>
        <Title.titleLeftSection titleName='Reviews' />
        <Title.titleRightSection titleName='View All' />
      </View>
      {/* End of reviews Title */}

      <View style={styles.reviewWrapper}>
        <View style={styles.reviewItemWrapper}>
          <View style={styles.reviewerInfoWrapper}>
            <Image style={styles.imageAvatar} source={{ uri: reviewers?.image }} />

            <View style={styles.reviewerItemInfo}>
              <Text style={styles.textName}>{reviewers?.name}</Text>

              <View style={styles.reviewTime}>
                <Image style={styles.imageClock} source={IMAGES.iconClock} />
                <Text style={styles.textDate}>{reviewers?.date}</Text>
              </View>
            </View>
          </View>
          {/* End of review Info */}

          <View style={styles.reviewItemRating}>
            <View style={styles.ratingTextWrapper}>
              {Boolean(rating) && <Text style={styles.ratingTextPoint}>{rating}</Text>}
              <Text style={styles.ratingText}>rating</Text>
            </View>

            <Image style={styles.imageStar} source={IMAGES.ratingStart} />
          </View>
          {/* End of review Rating */}
        </View>

        <View style={styles.reviewItemContent}>
          {Boolean(comment) && (
            <MoreLessText styleShowMoreText={styles.textContent} numberOfLines={2}>
              {comment}
            </MoreLessText>
          )}
        </View>
        {/* End of review Content */}
      </View>
    </View>
  );
};

export default memo(Reviews, isEqual);
