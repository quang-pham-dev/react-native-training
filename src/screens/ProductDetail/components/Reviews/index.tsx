import React, {memo} from 'react'

// LIBS
import isEqual from 'react-fast-compare'
import {s, vs} from 'react-native-size-matters/extend'

// Components
import MoreLessText from '@components/LessMoreText'

// Types
import {ReviewerProps} from '@model-types'

// Themes
import {Icons, Metrics} from '@themes'

// Styles
import ViewStyled from '@components/View/View.styles'
import FlexStyled from '@components/Flex/Flex.styles'
import PStyled from '@components/Paragraph/P.styles'
import IconStyled from '@components/Icon/Icon.styles'
import {Image} from '@components/Image/Image.styles'

// Constants
import {PARAGRAPH_TYPE} from '@constants'

type ReviewsProps = {
  comment: string
  rating: string
  reviewers: ReviewerProps
}

const Reviews = ({comment, rating, reviewers}: ReviewsProps) => {
  const {name, date, image} = reviewers || []

  return (
    <ViewStyled.Custom mTop={s(Metrics.margin.medium)}>
      <FlexStyled.FlexSpaceBetweenCenter>
        <PStyled.Base type={PARAGRAPH_TYPE.PRODUCT_DETAIL_TITLE_SECTION}>
          Reviews
        </PStyled.Base>
        <PStyled.Base type={PARAGRAPH_TYPE.PRODUCT_DETAIL_REVIEWS_VIEW_ALL}>
          View All
        </PStyled.Base>
      </FlexStyled.FlexSpaceBetweenCenter>
      {/* End of reviews Title */}

      <ViewStyled.Custom pTop={s(Metrics.padding.medium)}>
        <FlexStyled.FlexSpaceBetweenCenter>
          <FlexStyled.RowCenterHorizontal>
            <Image.Normal
              alignSelf="center"
              height={s(40)}
              width={s(40)}
              source={{
                uri: image,
              }}
              borderRadius={vs(40)}
              resizeMode="cover"
              mRight={s(Metrics.padding.custom)}
            />

            <FlexStyled.Default>
              <PStyled.Base
                type={PARAGRAPH_TYPE.PRODUCT_DETAIL_REVIEWS_USERNAME}
                mBottom={vs(8)}>
                {name}
              </PStyled.Base>

              <FlexStyled.RowCenter>
                <IconStyled source={Icons.clock} />
                <PStyled.Base
                  type={PARAGRAPH_TYPE.PRODUCT_DETAIL_REVIEWS_DATE_TIME}>
                  {date}
                </PStyled.Base>
              </FlexStyled.RowCenter>
            </FlexStyled.Default>
          </FlexStyled.RowCenterHorizontal>
          {/* End of review Info */}

          <FlexStyled.Default>
            <FlexStyled.RowCenterHorizontal>
              {!!rating && (
                <PStyled.Base
                  type={PARAGRAPH_TYPE.PRODUCT_DETAIL_REVIEWS_USERNAME}
                  mRight={s(Metrics.margin.tiny)}>
                  {rating}
                </PStyled.Base>
              )}
              <PStyled.Base
                type={PARAGRAPH_TYPE.PRODUCT_DETAIL_REVIEWS_DATE_TIME}>
                rating
              </PStyled.Base>
            </FlexStyled.RowCenterHorizontal>

            <IconStyled source={Icons.ratingStart} />
          </FlexStyled.Default>
          {/* End of review Rating */}
        </FlexStyled.FlexSpaceBetweenCenter>

        <ViewStyled.Custom mTop={vs(10)}>
          {Boolean(comment) && (
            <MoreLessText numberOfLines={2}>{comment}</MoreLessText>
          )}
        </ViewStyled.Custom>
        {/* End of review Content */}
      </ViewStyled.Custom>
    </ViewStyled.Custom>
  )
}

export default memo(Reviews, isEqual)
