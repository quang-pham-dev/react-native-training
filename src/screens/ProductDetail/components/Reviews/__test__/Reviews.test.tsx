import React from 'react'

// LIBS
import renderer from 'react-test-renderer'

// Components
import Reviews from '../index'

// Mock data
import {product} from '@constants'

const props = {
  comment: product.comment,
  rating: product.rating,
  reviewers: product.reviewers,
}
describe('Product detail reviewers', () => {
  const tree = renderer.create(<Reviews {...props} />)

  test('should render correctly', () => {
    expect(tree).toMatchSnapshot()
  })
})
