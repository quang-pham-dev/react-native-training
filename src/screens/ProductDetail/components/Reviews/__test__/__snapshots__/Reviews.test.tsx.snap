// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product detail reviewers should render correctly 1`] = `
<View
  mTop={15}
  style={
    Array [
      Object {
        "marginTop": 15,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1D1E20",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 17,
            "lineHeight": 18.7,
            "textAlign": "left",
          },
        ]
      }
      type="productDetailTitleSection"
    >
      Reviews
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#3D93F8",
            "fontFamily": "Inter-Regular",
            "fontSize": 13,
            "lineHeight": 14.3,
            "textAlign": "left",
          },
        ]
      }
      type="productDetailReviewsReviewAll"
    >
      View All
    </Text>
  </View>
  <View
    pTop={15}
    style={
      Array [
        Object {
          "paddingTop": 15,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
          ]
        }
      >
        <Image
          alignSelf="center"
          borderRadius={40}
          height={40}
          mRight={10}
          resizeMode="cover"
          source={
            Object {
              "uri": "https://images.unsplash.com/photo-1609505848912-b7c3b8b4beda?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=765&q=80",
            }
          }
          style={
            Array [
              Object {
                "alignSelf": "center",
                "borderRadius": 40,
                "height": 40,
                "marginRight": 10,
                "resizeMode": "cover",
                "width": 40,
              },
            ]
          }
          width={40}
        />
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            mBottom={8}
            style={
              Array [
                Object {
                  "color": "#1D1E20",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 15,
                  "lineHeight": 16.5,
                  "marginBottom": 8,
                  "textAlign": "left",
                },
              ]
            }
            type="productDetailReviewsUsername"
          >
            John Doe
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/assets/icons/clock.webp",
                }
              }
              style={
                Array [
                  Object {},
                ]
              }
            />
            <Text
              style={
                Array [
                  Object {
                    "color": "#8F959E",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 11,
                    "lineHeight": 11,
                    "textAlign": "left",
                  },
                ]
              }
              type="productDetailReviewsDateTime"
            >
              13, Sep, 2020
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              },
            ]
          }
        >
          <Text
            mRight={5}
            style={
              Array [
                Object {
                  "color": "#1D1E20",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 15,
                  "lineHeight": 16.5,
                  "marginRight": 5,
                  "textAlign": "left",
                },
              ]
            }
            type="productDetailReviewsUsername"
          >
            4.5
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#8F959E",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 11,
                  "lineHeight": 11,
                  "textAlign": "left",
                },
              ]
            }
            type="productDetailReviewsDateTime"
          >
            rating
          </Text>
        </View>
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/icons/start-rate.webp",
            }
          }
          style={
            Array [
              Object {},
            ]
          }
        />
      </View>
    </View>
    <View
      mTop={10}
      style={
        Array [
          Object {
            "marginTop": 10,
            "width": "100%",
          },
        ]
      }
    >
      <Text
        onTextLayout={[Function]}
        style={
          Array [
            Object {
              "color": "#8F959E",
              "fontFamily": "Inter-Regular",
              "fontSize": 15,
              "lineHeight": 21,
            },
          ]
        }
      >
        lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam,adipisicing elit. Quisquam, quia.
      </Text>
    </View>
  </View>
</View>
`;
